src/stan/model/indexing/rvalue_cl.hpp

```diff
@@ -0,0 +1,395 @@
+inline auto cl_row_index(index_min_max i, int rows, const char* name) {
+  math::check_range("min_max indexing min", name, rows, i.min_);
+  math::check_range("min_max indexing max", name, rows, i.max_);
+  if (i.min_ <= i.max_) {
+    return 1 * math::row_index(i.max_ - (i.min_ - 1), -1) + (i.min_ - 1);
+  } else {
+    return -1 * math::row_index(i.min_ - (i.max_ - 1), -1) + (i.min_ - 1);
+  }
+}
+
+inline auto cl_col_index(index_uni i, int cols, const char* name) {
+  math::check_range("uni indexing", name, cols, i.n_);
+  return math::constant(i.n_ - 1, -1, 1);
+}
+inline auto cl_col_index(const math::matrix_cl<int>& i, int cols,
+                         const char* name) {
+  return math::colwise_broadcast(math::transpose(i - 1));
+}
+inline auto cl_col_index(index_omni i, int cols, const char* name) {
+  return math::col_index(-1, cols);
+}
+inline auto cl_col_index(index_min i, int cols, const char* name) {
+  stan::math::check_range("min indexing", name, cols, i.min_);
+  return math::col_index(-1, cols - (i.min_ - 1)) + (i.min_ - 1);
+}
+inline auto cl_col_index(index_max i, int cols, const char* name) {
+  stan::math::check_range("max indexing", name, cols, i.max_);
+  return math::col_index(-1, i.max_);
+}
+inline auto cl_col_index(index_min_max i, int cols, const char* name) {
+  math::check_range("min_max indexing min", name, cols, i.min_);
+  math::check_range("min_max indexing max", name, cols, i.max_);
+  if (i.min_ <= i.max_) {
+    return 1 * math::col_index(-1, i.max_ - (i.min_ - 1)) + (i.min_ - 1);
+  } else {
+    return -1 * math::col_index(-1, i.min_ - (i.max_ - 1)) + (i.min_ - 1);
+  }
+}
+
+template <typename Index>
+inline std::tuple<> index_check(Index i, const char* name, int dim) {
+  return std::make_tuple();
+}
+inline auto index_check(const math::matrix_cl<int>& i, const char* name,
+                        int dim) {
+  return std::make_tuple(std::make_pair(
+      math::check_cl("multi-indexing", name, i, "within range"),
+      static_cast<int>(stan::error_index::value) <= i
+          && i < dim + static_cast<int>(stan::error_index::value)));
+}
+}  // namespace internal
+
+// prim
+/**
+ * Index a prim kernel generator expression with one index.
+ *
+ * @tparam Expr type of the exoression
+ * @tparam RowIndex type of index
+ * @param expr a prim kernel generator expression to index
+ * @param name string form of expression being evaluated
+ * @param row_index index
+ * @return result of indexing
+ */
+template <typename Expr, typename RowIndex,
+          require_all_kernel_expressions_and_none_scalar_t<Expr>* = nullptr>
+inline auto rvalue(Expr&& expr, const char* name, const RowIndex& row_index) {
+  auto checks = internal::index_check(row_index, name, expr.rows());
+  try {
+    math::index_apply<std::tuple_size<decltype(checks)>::value>(
+        [&checks](auto... Is) {
+          math::results(std::get<Is>(checks).first...)
+              = math::expressions(std::get<Is>(checks).second...);
+        });
```

Comment:
Is there a reason to use `index_apply` over `apply` here? i.e.

```suggestion
    math::apply([](const auto& check) {
          math::results(check.first...) = math::expressions(check.second...);
    }, checks);
```

It looks like you use this pattern a lot, would be good to add an internal function that just does this ala

```cpp
template <typename T>
void index_check_cl(const T& checks) {
    math::apply([](const auto& check) {
          math::results(check.first...) = math::expressions(check.second...);
    }, checks);
}
```

Reply:
Actually this is way more complicated than it needs to be. Fixed.